pinot-core/src/main/java/org/apache/pinot/core/operator/filter/FilterOperatorUtils.java

```diff
@@ -28,9 +28,218 @@
 
 
 public class FilterOperatorUtils {
+
+  private static Implementation _instance = new DefaultImplementation();
+
   private FilterOperatorUtils() {
   }
 
+  public static void setImplementation(Implementation newImplementation) {
+    _instance = newImplementation;
+  }
+
+  public interface Implementation {
+
+    /**
+     * Returns the leaf filter operator (i.e. not {@link AndFilterOperator} or {@link OrFilterOperator}).
+     */
+    BaseFilterOperator getLeafFilterOperator(PredicateEvaluator predicateEvaluator, DataSource dataSource,
+        int numDocs, boolean nullHandlingEnabled);
+
+    /**
+     * Returns the AND filter operator or equivalent filter operator.
+     */
+    BaseFilterOperator getAndFilterOperator(QueryContext queryContext,
+        List<BaseFilterOperator> filterOperators, int numDocs);
+
+    /**
+     * Returns the OR filter operator or equivalent filter operator.
+     */
+    BaseFilterOperator getOrFilterOperator(QueryContext queryContext,
+        List<BaseFilterOperator> filterOperators, int numDocs);
+
+    /**
+     * Returns the NOT filter operator or equivalent filter operator.
+     */
+    BaseFilterOperator getNotFilterOperator(QueryContext queryContext, BaseFilterOperator filterOperator,
+        int numDocs);
+
+    /**
+     * For AND filter operator, reorders its child filter operators based on the their cost and puts the ones with
+     * inverted index first in order to reduce the number of documents to be processed.
+     * <p>Special filter operators such as {@link MatchAllFilterOperator} and {@link EmptyFilterOperator} should be
+     * removed from the list before calling this method.
+     */
+    void reorderAndFilterChildOperators(QueryContext queryContext,
```

Comment:
This method should not be part of the interface because it is implementation detail for `getAndFilterOperator()`